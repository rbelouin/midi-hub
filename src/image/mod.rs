#[derive(Copy, Clone, Debug, PartialEq)]
pub struct Pixel {
    r: u8,
    g: u8,
    b: u8,
}

pub fn compress_64(width: u16, height: u16, pixels: Vec<Pixel>) -> Result<Vec<Pixel>, String> {
    if width * height != pixels.len() as u16 {
        return Err(format!("Number of pixels ({}) not matching width ({}) multiplied by height ({})", pixels.len(), width, height));
    }

    let mut output = vec![Pixel { r: 0, g: 0, b: 0 }; 64];
    for new_y in 0..8 {
        for new_x in 0..8 {
            let mut section = vec![];
            for y in (new_y * height / 8)..((new_y + 1) * height / 8) {
                for x in (new_x * width / 8)..((new_x + 1) * width / 8) {
                    section.push(pixels[(height * y + x) as usize]);
                }
            }
            output[(new_y * 8 + new_x) as usize] = compress(section);
        }
    }

    return Ok(output);
}

pub fn compress(pixels: Vec<Pixel>) -> Pixel {
    if pixels.is_empty() {
        return Pixel { r: 0, g: 0, b: 0 };
    }

    let mut r = 0u32;
    let mut g = 0u32;
    let mut b = 0u32;

    for i in 0..pixels.len() {
        r += pixels[i].r as u32;
        g += pixels[i].g as u32;
        b += pixels[i].b as u32;
    }

    let len = pixels.len() as u32;

    return Pixel {
        r: (r / len) as u8,
        g: (g / len) as u8,
        b: (b / len) as u8,
    };
}

#[cfg(test)]
mod tests {
    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::*;

    #[test]
    fn test_compress_64_invalid_input_should_return_error() {
        let result = compress_64(16, 16, vec![]);
        assert!(result.is_err(), "result should be an error, got {:?}", result);
    }

    #[test]
    fn test_compress_64_valid_input_should_compress_picture() {
        // black
        let b = Pixel { r: 0, g: 0, b: 0 };
        // white
        let w = Pixel { r: 255, g: 255, b: 255 };
        // grey
        let g = Pixel { r: 127, g: 127, b: 127 };

        let input = vec![
            b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            w, w, w, w, w, w, w, w, w, w, w, w, w, w, w, w,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            w, w, w, w, w, w, w, w, w, w, w, w, w, w, w, w,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            w, w, w, w, w, w, w, w, w, w, w, w, w, w, w, w,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            w, w, w, w, w, w, w, w, w, w, w, w, w, w, w, w,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
        ];

        let expected_output = vec![
            g, b, g, b, g, b, g, b,
            w, g, w, g, w, g, w, g,
            g, b, g, b, g, b, g, b,
            w, g, w, g, w, g, w, g,
            g, b, g, b, g, b, g, b,
            w, g, w, g, w, g, w, g,
            g, b, g, b, g, b, g, b,
            w, g, w, g, w, g, w, g,
        ];

        assert_eq!(compress_64(16, 16, input), Ok(expected_output));
    }

    #[test]
    fn test_compress_no_pixels_should_return_black() {
        assert_eq!(compress(vec![]), Pixel { r: 0, g: 0, b: 0 });
    }

    #[test]
    fn test_compress_monochrome_pixels_return_same_pixel() {
        let count = rand::random::<usize>() % 1023 + 1;
        let pixel = Pixel {
            r: rand::random::<u8>(),
            g: rand::random::<u8>(),
            b: rand::random::<u8>(),
        };
        let pixels = vec![pixel; count];

        assert_eq!(compress(pixels), pixel);
    }

    #[test]
    fn test_compress_shades_of_red_return_shade_of_red() {
        let count = rand::random::<usize>() % 1023 + 1;
        let mut pixels = vec![Pixel { r: 0, g: 0, b: 0 }; count];
        pixels[0].r = 30;
        for i in 1..pixels.len() {
            pixels[i].r = rand::random::<u8>();
        }

        let compressed_pixel = compress(pixels);
        assert!(compressed_pixel.r > 0, "red channel should be greater than zero, got {}", compressed_pixel.r);
        assert_eq!(compressed_pixel.g, 0);
        assert_eq!(compressed_pixel.b, 0);
    }

    #[test]
    fn test_compress_shades_of_green_return_shade_of_green() {
        let count = rand::random::<usize>() % 1023 + 1;
        let mut pixels = vec![Pixel { r: 0, g: 0, b: 0 }; count];
        pixels[0].g = 30;
        for i in 1..pixels.len() {
            pixels[i].g = rand::random::<u8>();
        }

        let compressed_pixel = compress(pixels);
        assert_eq!(compressed_pixel.r, 0);
        assert!(compressed_pixel.g > 0, "green channel should be greater than zero, got {}", compressed_pixel.g);
        assert_eq!(compressed_pixel.b, 0);
    }

    #[test]
    fn test_compress_shades_of_blue_return_shade_of_blue() {
        let count = rand::random::<usize>() % 1023 + 1;
        let mut pixels = vec![Pixel { r: 0, g: 0, b: 0 }; count];
        pixels[0].b = 30;
        for i in 1..pixels.len() {
            pixels[i].b = rand::random::<u8>();
        }

        let compressed_pixel = compress(pixels);
        assert_eq!(compressed_pixel.r, 0);
        assert_eq!(compressed_pixel.g, 0);
        assert!(compressed_pixel.b > 0, "blue channel should be greater than zero, got {}", compressed_pixel.b);
    }

    #[test]
    fn test_compress_return_between_extreme_values() {
        let mut pixels = vec![Pixel { r: 0, g: 0, b: 0 }; 2];

        let mut min_r = u8::MAX;
        let mut max_r = u8::MIN;
        let mut min_g = u8::MAX;
        let mut max_g = u8::MIN;
        let mut min_b = u8::MAX;
        let mut max_b = u8::MIN;

        for i in 0..pixels.len() {
            pixels[i].r = rand::random::<u8>();
            pixels[i].g = rand::random::<u8>();
            pixels[i].b = rand::random::<u8>();

            if pixels[i].r < min_r {
                min_r = pixels[i].r;
            }

            if pixels[i].r > max_r {
                max_r = pixels[i].r;
            }

            if pixels[i].g < min_g {
                min_g = pixels[i].g;
            }

            if pixels[i].g > max_g {
                max_g = pixels[i].g;
            }

            if pixels[i].b < min_b {
                min_b = pixels[i].b;
            }

            if pixels[i].b > max_b {
                max_b = pixels[i].b;
            }
        }

        let compressed_pixel = compress(pixels);
        assert!(min_r <= compressed_pixel.r && max_r >= compressed_pixel.r, "red channel should be in [{}; {}], got {}", min_r, max_r, compressed_pixel.r);
        assert!(min_g <= compressed_pixel.g && max_g >= compressed_pixel.g, "green channel should be in [{}; {}], got {}", min_g, max_g, compressed_pixel.g);
        assert!(min_b <= compressed_pixel.b && max_b >= compressed_pixel.b, "blue channel should be in [{}; {}], got {}", min_b, max_b, compressed_pixel.b);
    }
}
