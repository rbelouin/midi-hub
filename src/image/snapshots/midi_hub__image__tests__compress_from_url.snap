---
source: src/image/mod.rs
assertion_line: 120
expression: result

---
Ok(
    [
        Pixel {
            r: 183,
            g: 183,
            b: 194,
        },
        Pixel {
            r: 176,
            g: 180,
            b: 184,
        },
        Pixel {
            r: 206,
            g: 194,
            b: 201,
        },
        Pixel {
            r: 208,
            g: 194,
            b: 202,
        },
        Pixel {
            r: 213,
            g: 199,
            b: 206,
        },
        Pixel {
            r: 208,
            g: 193,
            b: 200,
        },
        Pixel {
            r: 201,
            g: 161,
            b: 168,
        },
        Pixel {
            r: 197,
            g: 151,
            b: 168,
        },
        Pixel {
            r: 151,
            g: 117,
            b: 116,
        },
        Pixel {
            r: 135,
            g: 119,
            b: 113,
        },
        Pixel {
            r: 167,
            g: 122,
            b: 121,
        },
        Pixel {
            r: 181,
            g: 131,
            b: 129,
        },
        Pixel {
            r: 163,
            g: 146,
            b: 153,
        },
        Pixel {
            r: 161,
            g: 139,
            b: 134,
        },
        Pixel {
            r: 184,
            g: 86,
            b: 86,
        },
        Pixel {
            r: 180,
            g: 106,
            b: 109,
        },
        Pixel {
            r: 160,
            g: 157,
            b: 170,
        },
        Pixel {
            r: 69,
            g: 70,
            b: 68,
        },
        Pixel {
            r: 150,
            g: 140,
            b: 157,
        },
        Pixel {
            r: 185,
            g: 96,
            b: 105,
        },
        Pixel {
            r: 117,
            g: 109,
            b: 119,
        },
        Pixel {
            r: 98,
            g: 97,
            b: 103,
        },
        Pixel {
            r: 175,
            g: 81,
            b: 89,
        },
        Pixel {
            r: 148,
            g: 121,
            b: 142,
        },
        Pixel {
            r: 139,
            g: 126,
            b: 126,
        },
        Pixel {
            r: 89,
            g: 79,
            b: 38,
        },
        Pixel {
            r: 140,
            g: 119,
            b: 132,
        },
        Pixel {
            r: 174,
            g: 78,
            b: 76,
        },
        Pixel {
            r: 86,
            g: 98,
            b: 126,
        },
        Pixel {
            r: 100,
            g: 111,
            b: 142,
        },
        Pixel {
            r: 39,
            g: 55,
            b: 64,
        },
        Pixel {
            r: 125,
            g: 124,
            b: 146,
        },
        Pixel {
            r: 64,
            g: 63,
            b: 56,
        },
        Pixel {
            r: 61,
            g: 58,
            b: 44,
        },
        Pixel {
            r: 81,
            g: 53,
            b: 47,
        },
        Pixel {
            r: 82,
            g: 53,
            b: 51,
        },
        Pixel {
            r: 60,
            g: 64,
            b: 77,
        },
        Pixel {
            r: 68,
            g: 73,
            b: 89,
        },
        Pixel {
            r: 31,
            g: 44,
            b: 48,
        },
        Pixel {
            r: 75,
            g: 73,
            b: 77,
        },
        Pixel {
            r: 24,
            g: 25,
            b: 20,
        },
        Pixel {
            r: 21,
            g: 21,
            b: 15,
        },
        Pixel {
            r: 31,
            g: 33,
            b: 27,
        },
        Pixel {
            r: 24,
            g: 25,
            b: 19,
        },
        Pixel {
            r: 34,
            g: 30,
            b: 24,
        },
        Pixel {
            r: 30,
            g: 28,
            b: 22,
        },
        Pixel {
            r: 41,
            g: 43,
            b: 36,
        },
        Pixel {
            r: 33,
            g: 30,
            b: 23,
        },
        Pixel {
            r: 103,
            g: 39,
            b: 31,
        },
        Pixel {
            r: 198,
            g: 82,
            b: 62,
        },
        Pixel {
            r: 224,
            g: 111,
            b: 81,
        },
        Pixel {
            r: 237,
            g: 118,
            b: 86,
        },
        Pixel {
            r: 240,
            g: 122,
            b: 88,
        },
        Pixel {
            r: 235,
            g: 111,
            b: 81,
        },
        Pixel {
            r: 216,
            g: 84,
            b: 64,
        },
        Pixel {
            r: 112,
            g: 41,
            b: 33,
        },
        Pixel {
            r: 64,
            g: 34,
            b: 27,
        },
        Pixel {
            r: 112,
            g: 41,
            b: 32,
        },
        Pixel {
            r: 97,
            g: 60,
            b: 44,
        },
        Pixel {
            r: 91,
            g: 51,
            b: 38,
        },
        Pixel {
            r: 96,
            g: 55,
            b: 40,
        },
        Pixel {
            r: 110,
            g: 56,
            b: 42,
        },
        Pixel {
            r: 134,
            g: 45,
            b: 36,
        },
        Pixel {
            r: 65,
            g: 34,
            b: 26,
        },
    ],
)
